File: script/robust_stats.py
"""
robust_stats.py
---------------
Generische robuste Statistik-Utilities für Finanz-Analysen
(Reverse-DCF-Basket, Peer-Vergleiche, etc.).

Python-Referenzimplementierung (identisch zur TypeScript-Version
in client/src/lib/robustStats.ts).

Methoden:
- Quantile R-7 (Excel PERCENTILE.INC / NumPy linear)
- Winsorisierung
- Winsorized Median
- Getrimmter Mittelwert (optional)

Keine externen Abhängigkeiten außer der Standardbibliothek.
"""

from __future__ import annotations
from typing import Sequence, List, Optional
import math
import unittest


def quantile_r7(data: Sequence[float], p: float) -> float:
    """
    Quantile nach Hyndman-Fan Typ 7 (R-7).
    Identisch mit Excel PERCENTILE.INC und NumPy method='linear'.
    """
    if not 0.0 <= p <= 1.0:
        raise ValueError("p muss im Intervall [0, 1] liegen")

    sorted_data = sorted(x for x in data if math.isfinite(x))
    n = len(sorted_data)
    if n == 0:
        raise ValueError("Keine endlichen Werte")
    if n == 1:
        return float(sorted_data[0])
    if p == 0.0:
        return float(sorted_data[0])
    if p == 1.0:
        return float(sorted_data[-1])

    h = p * (n - 1)
    h_floor = math.floor(h)
    h_ceil = math.ceil(h)
    frac = h - h_floor

    return float(sorted_data[h_floor] * (1.0 - frac) + sorted_data[h_ceil] * frac)


def winsorize(
    data: Sequence[float],
    lower: float = 0.05,
    upper: float = 0.95,
) -> List[float]:
    """
    Winsorisierung: Extreme Werte werden auf die Quantile gesetzt.
    Beobachtungen bleiben erhalten (n ändert sich nicht).
    """
    if not 0.0 <= lower < upper <= 1.0:
        raise ValueError("Es muss 0 <= lower < upper <= 1 gelten")

    clean = [x for x in data if math.isfinite(x)]
    if len(clean) < 4:
        return clean[:]  # zu klein → keine Winsorisierung

    q_low = quantile_r7(clean, lower)
    q_high = quantile_r7(clean, upper)

    return [max(q_low, min(q_high, x)) for x in clean]


def winsorized_median(
    data: Sequence[float],
    lower: float = 0.05,
    upper: float = 0.95,
) -> Optional[float]:
    """
    Empfohlene Aggregationsfunktion für Peer-Baskets:
    Median der winsorisierten Werte.
    """
    w = winsorize(data, lower, upper)
    if not w:
        return None
    s = sorted(w)
    mid = len(s) // 2
    if len(s) % 2 == 0:
        return (s[mid - 1] + s[mid]) / 2.0
    return float(s[mid])


def trimmed_mean(
    data: Sequence[float],
    proportiontocut: float = 0.05,
) -> Optional[float]:
    """
    Getrimmter Mittelwert (optional).
    Entfernt den Anteil proportiontocut unten und oben.
    Bei kleinen n oft wirkungslos → Winsorisierung bevorzugen.
    """
    clean = sorted(x for x in data if math.isfinite(x))
    n = len(clean)
    if n == 0:
        return None

    k = int(math.floor(proportiontocut * n))
    if 2 * k >= n:
        return None  # zu aggressiv

    trimmed = clean[k : n - k]
    return sum(trimmed) / len(trimmed)


def compute_basket_growth(
    revenue_cagrs: Sequence[float],
    eps_cagrs: Sequence[float],
    lower: float = 0.05,
    upper: float = 0.95,
) -> Optional[float]:
    """
    g_basket = 0.6 * winsorized_median(Revenue-CAGR) + 0.4 * winsorized_median(EPS-CAGR)
    """
    g_rev = winsorized_median(revenue_cagrs, lower, upper)
    g_eps = winsorized_median(eps_cagrs, lower, upper)

    if g_rev is None and g_eps is None:
        return None
    if g_rev is None:
        return g_eps
    if g_eps is None:
        return g_rev
    return 0.6 * g_rev + 0.4 * g_eps


# ---------------------------------------------------------------------------
# Unit-Tests
# ---------------------------------------------------------------------------

class TestRobustStats(unittest.TestCase):
    DATA = [-18.2, 3.1, 8.4, 11.0, 13.7, 15.9, 21.4, 94.6]

    def test_quantile_r7_excel_compatibility(self):
        q05 = quantile_r7(self.DATA, 0.05)
        q95 = quantile_r7(self.DATA, 0.95)
        self.assertAlmostEqual(q05, -10.745, places=3)
        self.assertAlmostEqual(q95, 68.98, places=2)

    def test_quantile_edge_cases(self):
        self.assertEqual(quantile_r7([5.0], 0.5), 5.0)
        self.assertEqual(quantile_r7(self.DATA, 0.0), -18.2)
        self.assertEqual(quantile_r7(self.DATA, 1.0), 94.6)

    def test_winsorize_5_95(self):
        w = winsorize(self.DATA, 0.05, 0.95)
        self.assertEqual(len(w), 8)
        self.assertAlmostEqual(min(w), -10.745, places=3)
        self.assertAlmostEqual(max(w), 68.98, places=2)

    def test_winsorized_median(self):
        med = winsorized_median(self.DATA)
        self.assertAlmostEqual(med, 12.35, places=2)

    def test_small_sample_no_winsorize(self):
        small = [1.0, 2.0, 100.0]
        w = winsorize(small)
        self.assertEqual(w, [1.0, 2.0, 100.0])

    def test_trimmed_mean(self):
        tm = trimmed_mean(self.DATA, 0.05)
        classic_mean = sum(self.DATA) / len(self.DATA)
        self.assertAlmostEqual(tm, classic_mean, places=5)

        tm2 = trimmed_mean(self.DATA, 0.125)
        self.assertAlmostEqual(tm2, 12.25, places=2)

    def test_compute_basket_growth(self):
        rev = [0.04, 0.09, 0.11, 0.14, 0.16, 0.22, 0.87, -0.12]
        eps = [0.03, 0.08, 0.10, 0.13, 0.15, 0.20, 0.70, -0.15]
        g = compute_basket_growth(rev, eps)
        self.assertIsNotNone(g)
        self.assertTrue(0 < g < 0.5)


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: client/src/lib/robustStats.ts
/**
 * robustStats.ts
 * --------------
 * Generische robuste Statistik-Utilities für Finanz-Analysen
 * (Reverse-DCF-Basket, Peer-Vergleiche, Sektor-Mediane etc.).
 *
 * Methoden:
 * - Quantile R-7 (Excel PERCENTILE.INC / NumPy linear) — generisch, kein Hardcode
 * - Winsorisierung (Extreme auf Quantile setzen, n bleibt erhalten)
 * - Winsorized Median (empfohlene Aggregation für Peer-Baskets)
 * - Getrimmter Mittelwert (optionale Alternative)
 *
 * Keine Ticker-/Sektor-Hardcodes. Alle Funktionen sind pure.
 *
 * Spezifikation: WORK_VALUECHAIN_SECTOR_ROTATION.md §5 + Gespräch 17.08.2026
 */

export type FiniteNumberArray = number[];

/**
 * Quantile nach Hyndman-Fan Typ 7 (R-7).
 * Identisch mit Excel PERCENTILE.INC und NumPy method='linear'.
 *
 * Formel: h = p * (n - 1)
 *         Q = sorted[⌊h⌋] * (1 - {h}) + sorted[⌈h⌉] * {h}
 */
export function quantileR7(data: number[], p: number): number {
  if (p < 0 || p > 1 || !Number.isFinite(p)) {
    throw new Error(`p must be in [0, 1], got ${p}`);
  }

  const sorted = data.filter((x) => Number.isFinite(x)).sort((a, b) => a - b);
  const n = sorted.length;

  if (n === 0) throw new Error("No finite values");
  if (n === 1) return sorted[0];
  if (p === 0) return sorted[0];
  if (p === 1) return sorted[n - 1];

  const h = p * (n - 1);
  const hFloor = Math.floor(h);
  const hCeil = Math.ceil(h);
  const frac = h - hFloor;

  return sorted[hFloor] * (1 - frac) + sorted[hCeil] * frac;
}

/**
 * Winsorisierung: Extreme Werte werden auf die Quantile gesetzt.
 * Beobachtungen bleiben erhalten (n ändert sich nicht).
 *
 * Default: 5 % / 95 % (empfohlen für Peer-Baskets n = 4–15).
 * Bei n < 4 wird unverändert zurückgegeben.
 */
export function winsorize(
  data: number[],
  lower = 0.05,
  upper = 0.95
): number[] {
  if (!(0 <= lower && lower < upper && upper <= 1)) {
    throw new Error(`Invalid bounds: lower=${lower}, upper=${upper}`);
  }

  const clean = data.filter((x) => Number.isFinite(x));
  if (clean.length < 4) return [...clean];

  const qLow = quantileR7(clean, lower);
  const qHigh = quantileR7(clean, upper);

  return clean.map((x) => Math.max(qLow, Math.min(qHigh, x)));
}

/**
 * Empfohlene Aggregationsfunktion für Reverse-DCF-Baskets und Peer-Vergleiche:
 * Median der winsorisierten Werte.
 *
 * @returns null wenn keine endlichen Werte vorhanden
 */
export function winsorizedMedian(
  data: number[],
  lower = 0.05,
  upper = 0.95
): number | null {
  const w = winsorize(data, lower, upper);
  if (w.length === 0) return null;

  const s = [...w].sort((a, b) => a - b);
  const mid = Math.floor(s.length / 2);

  if (s.length % 2 === 0) {
    return (s[mid - 1] + s[mid]) / 2;
  }
  return s[mid];
}

/**
 * Getrimmter Mittelwert (optionale Alternative).
 * Entfernt den Anteil `proportionToCut` unten und oben.
 * Bei kleinen n oft wirkungslos → Winsorisierung bevorzugen.
 */
export function trimmedMean(
  data: number[],
  proportionToCut = 0.05
): number | null {
  const clean = data.filter((x) => Number.isFinite(x)).sort((a, b) => a - b);
  const n = clean.length;
  if (n === 0) return null;

  const k = Math.floor(proportionToCut * n);
  if (2 * k >= n) return null; // zu aggressiv

  const trimmed = clean.slice(k, n - k);
  return trimmed.reduce((sum, x) => sum + x, 0) / trimmed.length;
}

/**
 * Convenience: berechnet g_basket für den Reverse-DCF-Vergleich.
 * Gewichtung: 60 % Revenue-CAGR + 40 % EPS-CAGR (winsorized Median).
 * Gibt null zurück, wenn einer der Inputs unbrauchbar ist.
 */
export function computeBasketGrowth(
  revenueCagrs: number[],
  epsCagrs: number[],
  lower = 0.05,
  upper = 0.95
): number | null {
  const gRev = winsorizedMedian(revenueCagrs, lower, upper);
  const gEps = winsorizedMedian(epsCagrs, lower, upper);

  if (gRev == null && gEps == null) return null;
  if (gRev == null) return gEps;
  if (gEps == null) return gRev;

  return 0.6 * gRev + 0.4 * gEps;
}
